torch/utils/_pytree.py

```diff
@@ -172,6 +176,26 @@ def _register_pytree_node(
     SUPPORTED_SERIALIZED_TYPES[typ] = serialize_node_def
     SERIALIZED_TYPE_TO_PYTHON_TYPE[type_fqn] = typ
 
+    if _register_cxx_pytree_node:
+        import torch
+
+        if torch._running_with_deploy():
+            return
```

Comment:
What is the problem with deploy?

Reply:
The `TorchpyTest.TaggingRace` test gets internal errors. I'm not sure if we can fix this by adding a thread lock here. Or we can change the import order for `torch._C`.

https://github.com/pytorch/pytorch/actions/runs/6456239485/job/17526163874?pr=109684#step:15:2360

```text
Traceback (most recent call last):
  File "<string>", line 49, in <module>
  File "/opt/conda/envs/py_3.10/lib/python3.10/site-packages/torch/__init__.py", line 236, in <module>
    from torch._C import *  # noqa: F403
RuntimeError: get_internals: could not successfully initialize the tstate TSS key!
unknown file: Failure
C++ exception with description "r == 0 INTERNAL ASSERT FAILED at "/var/lib/jenkins/multipy/multipy/runtime/interpreter/interpreter_impl.cpp":691, please report a bug to PyTorch. 
Exception raised from newInterpreterImpl at /var/lib/jenkins/multipy/multipy/runtime/interpreter/interpreter_impl.cpp:691 (most recent call first):
frame #0: c10::Error::Error(c10::SourceLocation, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >) + 0x6c (0x7f39f423ee7c in /opt/conda/envs/py_3.10/lib/python3.10/site-packages/torch/lib/libc10.so)
frame #1: c10::detail::torchCheckFail(char const*, char const*, unsigned int, char const*) + 0x84 (0x7f39f41f0174 in /opt/conda/envs/py_3.10/lib/python3.10/site-packages/torch/lib/libc10.so)
frame #2: c10::detail::deprecated_AT_ASSERT() + 0 (0x7f381e541a42 in /tmp/multipy_interpreterkpkukY)
frame #3: newInterpreterImpl + 0x98 (0x7f381e53f51c in /tmp/multipy_interpreterkpkukY)
frame #4: torch::deploy::Interpreter::Interpreter(torch::deploy::InterpreterManager*, std::shared_ptr<torch::deploy::Environment>) + 0x322 (0x55706d94f3fc in ./multipy/runtime/build/test_deploy)
frame #5: torch::deploy::Interpreter::Interpreter(std::shared_ptr<torch::deploy::Environment>) + 0x4c (0x55706d95101e in ./multipy/runtime/build/test_deploy)
frame #6: void __gnu_cxx::new_allocator<torch::deploy::Interpreter>::construct<torch::deploy::Interpreter, std::shared_ptr<torch::deploy::Environment>&>(torch::deploy::Interpreter*, std::shared_ptr<torch::deploy::Environment>&) + 0x6d (0x55706d954b05 in ./multipy/runtime/build/test_deploy)
frame #7: void std::allocator_traits<std::allocator<torch::deploy::Interpreter> >::construct<torch::deploy::Interpreter, std::shared_ptr<torch::deploy::Environment>&>(std::allocator<torch::deploy::Interpreter>&, torch::deploy::Interpreter*, std::shared_ptr<torch::deploy::Environment>&) + 0x3a (0x55706d953520 in ./multipy/runtime/build/test_deploy)
frame #8: torch::deploy::Interpreter& std::vector<torch::deploy::Interpreter, std::allocator<torch::deploy::Interpreter> >::emplace_back<std::shared_ptr<torch::deploy::Environment>&>(std::shared_ptr<torch::deploy::Environment>&) + 0x50 (0x55706d95216c in ./multipy/runtime/build/test_deploy)
frame #9: torch::deploy::InterpreterManager::InterpreterManager(unsigned long, std::shared_ptr<torch::deploy::Environment>) + 0x1d9 (0x55706d94ca29 in ./multipy/runtime/build/test_deploy)
frame #10: TorchpyTest_TaggingRace_Test::TestBody() + 0x7c (0x55706d8cec5c in ./multipy/runtime/build/test_deploy)
frame #11: void testing::internal::HandleSehExceptionsInMethodIfSupported<testing::Test, void>(testing::Test*, void (testing::Test::*)(), char const*) + 0x69 (0x55706d93a566 in ./multipy/runtime/build/test_deploy)
frame #12: void testing::internal::HandleExceptionsInMethodIfSupported<testing::Test, void>(testing::Test*, void (testing::Test::*)(), char const*) + 0x5e (0x55706d932ae7 in ./multipy/runtime/build/test_deploy)
frame #13: testing::Test::Run() + 0xf2 (0x55706d90acb4 in ./multipy/runtime/build/test_deploy)
frame #14: testing::TestInfo::Run() + 0x113 (0x55706d90b6c9 in ./multipy/runtime/build/test_deploy)
frame #15: testing::TestSuite::Run() + 0x131 (0x55706d90bfcb in ./multipy/runtime/build/test_deploy)
frame #16: testing::internal::UnitTestImpl::RunAllTests() + 0x431 (0x55706d918e55 in ./multipy/runtime/build/test_deploy)
frame #17: bool testing::internal::HandleSehExceptionsInMethodIfSupported<testing::internal::UnitTestImpl, bool>(testing::internal::UnitTestImpl*, bool (testing::internal::UnitTestImpl::*)(), char const*) + 0x69 (0x55706d93b627 in ./multipy/runtime/build/test_deploy)
frame #18: bool testing::internal::HandleExceptionsInMethodIfSupported<testing::internal::UnitTestImpl, bool>(testing::internal::UnitTestImpl*, bool (testing::internal::UnitTestImpl::*)(), char const*) + 0x5e (0x55706d933dad in ./multipy/runtime/build/test_deploy)
frame #19: testing::UnitTest::Run() + 0xcd (0x55706d91756f in ./multipy/runtime/build/test_deploy)
frame #20: RUN_ALL_TESTS() + 0x15 (0x55706d8d93f0 in ./multipy/runtime/build/test_deploy)
frame #21: main + 0x9d (0x55706d8d348d in ./multipy/runtime/build/test_deploy)
frame #22: __libc_start_main + 0xf3 (0x7f39e5779083 in /usr/lib/x86_64-linux-gnu/libc.so.6)
frame #23: _start + 0x2e (0x55706d8c855e in ./multipy/runtime/build/test_deploy)
" thrown in the test body.
[  FAILED  ] TorchpyTest.TaggingRace (14370 ms)
```

https://github.com/pytorch/multipy/blob/765dfef74b385532a2f707c20af8407d7f91cf70/multipy/runtime/interpreter/interpreter_impl.cpp#L690-L691

Reply:
Further discussion moved to PR:

- #112477

I add a warning that C++ pytree is not available under `torchdeploy`.